### PR TITLE
Performance Improvements

### LINKED ISSUE #334: Performance: highlights are slow in large datasets
See #170

In general, app performance is very poor for large datasets.

### PR DESCRIPTION
**Pull request by NateLanza**
### Does this PR close any open issues?
Closes #334

### Give a longer description of what this PR addresses and why it's needed
Improves performance across a variety of actions by reducing unnecessary re-renders. Includes:
- Adding WhyDidYouRender plugin to dev mode for debugging re-renders
- Wrapping the VegaLite element in the attribute density plots with a memoized component that only depends on strictly necessary props
- Preventing unnecessary re-renders during sorting, aggregation, hovering, selection, and visible attribute changes
- Using Recoil state in App to memoize rows
- Replacing VERY costly `getValuesFromRow` function in `AttributeBar` with a memoized Recoil selector
- Memoizing props fed to `VegaLite` component in element view sidebar
- Not rendering `ElementSidebar` component unless it is open

### Provide pictures/videos of the behavior before and after these changes (optional)
No visible changes to behavior other than speedup

### Have you added or updated relevant tests?
- [ ] Yes
- [x] No changes are needed

### Have you added or updated relevant documentation?
- [ ] Yes
- [x] No changes are needed

### CODE REVIEW

File: packages/upset/src/components/Root.tsx [hunk 1]
```diff
@@ -169,7 +169,7 @@ export const Root: FC<Props> = ({
         </SvgBase>
       </div>
       <ContextMenu />
-      {elementSidebar && <ElementSidebar open={elementSidebar.open} close={elementSidebar.close} />}
+      {elementSidebar && elementSidebar.open && <ElementSidebar open={elementSidebar.open} close={elementSidebar.close} />}
```
Comment (JackWilb) [inline packages/upset/src/components/Root.tsx:172]: Can we remove the open prop if we're using this render method?

File: packages/upset/src/components/ElementView/ElementVisualization.tsx [hunk 2]
```diff
@@ -34,14 +35,17 @@ export const ElementVisualization = () => {
...
+   * Internal State
    */
 
   const draftSelection = useRef(elementSelection);
   const vegaSpec = useMemo(
     () => generateVega(scatterplots, histograms, selectColor, elementSelection),
     [scatterplots, histograms, selectColor, elementSelection],
   );
+  const data = useMemo(() => ({
+    elements: Object.values(JSON.parse(JSON.stringify(items))),
```
Comment (JackWilb) [inline packages/upset/src/components/ElementView/ElementVisualization.tsx:47]: Would a `structuredClone` here be better than the parse/stringify hack?

File: packages/app/src/components/Header/index.tsx [hunk 3]
```diff
@@ -34,6 +35,7 @@ const Header = ({ data }: { data: any }) => {
   const setLoading = useSetRecoilState(loadingAtom);
   
   const { provenance } = useContext(ProvenanceContext);
+  const rows = useRecoilValue(rowsSelector(provenance.getState()))
```
Comment (JakeWags) [inline packages/app/src/components/Header/index.tsx:38]: Same here as what Jack brought up with provenance reactivity

File: packages/app/tsconfig.json [hunk 4]
```diff
@@ -1,7 +1,7 @@
 {
   "compilerOptions": {
     "target": "es5",
-    "types": ["@emotion/react/types/css-prop"],
+    "types": ["@emotion/react/types/css-prop", "node"],
```
Comment (JakeWags) [inline packages/app/tsconfig.json:4]: Why is this needed?

File: packages/upset/src/components/Columns/Attribute/AttributePlots/MemoizedDensityVega.tsx [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+        prevProps.values.length > 0
+          ? prevProps.values[0] === nextProps.values[0]
+            && prevProps.values[prevProps.values.length - 1] === nextProps.values[nextProps.values.length - 1]
+          : true
+      )
+      && prevProps.height === nextProps.height
+      && prevProps.fillColor === nextProps.fillColor
+      && prevProps.min === nextProps.min
+      && prevProps.max === nextProps.max,
+);
```
Comment (JakeWags) [inline packages/upset/src/components/Columns/Attribute/AttributePlots/MemoizedDensityVega.tsx:62]: Maybe worth writing an isEqual or similar function for this (and the other HOC in this PR). It is functionally the same thing, but a bit more clear imo.

File: packages/upset/src/components/ElementView/ElementVisualization.tsx (hunk 2 above)
Comment (JakeWags) [reply packages/upset/src/components/ElementView/ElementVisualization.tsx:47]: +1

**Review by JakeWags — APPROVED** [review]
Looks great. A few changes for clarity, but not blocks to merging.

File: packages/app/tsconfig.json (hunk 4 above)
Comment (JackWilb) [reply packages/app/tsconfig.json:4]: For structuredClone. I think this is included by default with newer versions of node, but if you're on 18(?) you need to add this.

**Review by JackWilb — APPROVED** [review]
